user-operator/src/main/java/io/strimzi/operator/user/operator/KafkaUserQuotasOperator.java

```diff
@@ -202,119 +92,105 @@ boolean isPathExist(String path)    {
-                String prod = config.getString("producer_byte_rate");
-                String cons = config.getString("consumer_byte_rate");
-                String perc = config.getString("request_percentage");
-
-                if (prod != null || cons != null || perc != null) {
-                    return true;
-                }
-            }
-        }
-        return false;
-    }
-
-    private boolean configJsonIsEmpty(JsonObject json) {
-        validateJsonVersion(json);
-        JsonObject config = json.getJsonObject("config");
-        return config.isEmpty();
+    boolean exists(String username) throws Exception {
+        return describeUserQuotas(username) != null;
     }
 
     /**
      * Delete the quotas for the given user.
      * It is not an error if the user doesn't exist, or doesn't currently have any quotas.
      *
      * @param username Name of the user
+     * @throws Exception when altering quotas fails
      */
-    public void delete(String username) {
-        String encodedUsername = encodeUsername(username);
-
-        byte[] data = zkClient.readData("/config/users/" + encodedUsername, true);
-
-        if (data != null)   {
+    public void delete(String username) throws Exception {
+        KafkaUserQuotas current = describeUserQuotas(username);
+        if (current != null) {
             log.debug("Deleting quotas for user {}", username);
-            JsonObject deleteJson = removeQuotasFromJsonUser(data);
-            if (configJsonIsEmpty(deleteJson)) {
-                zkClient.deleteRecursive("/config/users/" + encodedUsername);
-                log.debug("User {} deleted from ZK store", username);
-            } else {
-                zkClient.writeData("/config/users/" + encodedUsername, deleteJson.toBuffer().getBytes());
-            }
-            notifyChanges(username);
+            current.setProducerByteRate(null);
+            current.setConsumerByteRate(null);
+            current.setRequestPercentage(null);
+            alterUserQuotas(username, toClientQuotaAlterationOps(current));
         } else {
             log.warn("Quotas for user {} already don't exist", username);
         }
     }
 
-    /**
-     * Deletes the quotas from existing JSON
-     *
-     * @param userConfig JSON string with existing userConfig configuration as byte[]
-     *
-     * @return  Returns the updated JSON without the quotas
-     */
-    protected JsonObject removeQuotasFromJsonUser(byte[] userConfig)   {
-        JsonObject json = new JsonObject(new String(userConfig, StandardCharsets.UTF_8));
+    protected void alterUserQuotas(String username, Set<ClientQuotaAlteration.Op> ops) throws Exception {
+        ClientQuotaEntity cqe = new ClientQuotaEntity(Collections.singletonMap(ClientQuotaEntity.USER, username));
+        ClientQuotaAlteration cqa = new ClientQuotaAlteration(cqe, ops);
+        try {
+            adminClient.alterClientQuotas(Collections.singleton(cqa)).all().get();
```

Comment:
Should this be done asynchronously? Or is this whole block in some `executeBlocking`?

Reply:
Ditto as above.